### PR TITLE
First completed draft of ID Generation post.

### PR DESCRIPTION
**Pull request by nathankleyn**
Also added author image, and post image and thumbnails. Added `.gitkeep`
file to the `_drafts` folder to stop it from disappearing every time we
publish our only draft.

### CODE REVIEW

File: _drafts/id-generation.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+excerpt: How to generate globally unique IDs in a distributed fashion with Redis and Lua.
+image: /assets/images/id-generation.jpg
+thumbnail: /assets/images/id-generation-thumbnail.jpg
+categories: engineering
+author: Nathan Kleyn
+---
+
+**TL;DR:** We are open-sourcing our distributed _k_-ordered ID generation project called "Icicle", which generates IDs using Lua scripting within distributed Redis hosts. You can [find more information in the GitHub repository](https://github.com/intenthq/icicle)!
+
+At Intent HQ, we generate a lot of data. Part of the task of capturing this data is the package it up into a common format and store it in a way it can be retrieved. Commonly, labelling this data in a way that facilitates retrival is a task handled by your database. However, in a big-data environment, we need a way to label these pieces of data with an ID in a distributed and scalable way independent of a database.
```
Comment (albertpastrana) [inline _drafts/id-generation.md]: `...capturing the data is the package it up...` -> `...capturing the data is TO package it up...`

`retrival` -> `retrieval`

File: _drafts/id-generation.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+Over the course of this article, we'll walk you through how we arrived at a solution for generating IDs in a distributed fashion using the Redis instances we already had deployed and in a format that allows _k_-ordering. We'll share with you the code for doing this so you can use this ID generation in your own projects!
+
+## What is an ID?
+
+An ID can be many things, but surely its most important definition is that of "uniqueness". Of all the properties, this is the only one that truly matters. If you label a piece of data with an ID, you want to be sure you don't later have abiguities when looking up that data again by the ID you had.
```
Comment (albertpastrana) [inline _drafts/id-generation.md]: `abiguities` -> `ambiguities`

File: _drafts/id-generation.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+IDs may have some other useful properties, however:
+
+* _Ordered:_ They may be ordered to allow comparison of items stamped with IDs.
+
+* _Meaningful:_ They may embed certain information about the point at which the ID was generated, for example the time (however inexact).
+
+* _Distributed:_ They may be generated by a machine without consulating any other machine (ie. they must be truly distributed).
```
Comment (philmes) [inline _drafts/id-generation.md]: consulating -> consulting

File: _drafts/id-generation.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+As a turns out, primarily, the reasons for doing this are deeply rooted in our inability to keep time across a distributed environment.
+
+## Time is hard
+
+Keeping time is actually a really hard problem to solve. At the core of the issue is the fact that computers suffer from "clock drift", where the clock is constantly skewing forwards or backwards away from the actual time. Perhaps even worse, [every computer skews at a different rate](https://en.wikipedia.org/wiki/Clock_skew).
+
+Thus, the concept of a "time oracle" is born. Time oracles dedicated servers normally equipped with an atomic clock or GPS device whose sole purpose is for keeping time in as precise a manner as possible. Time oracle servers are used frequently in environments where super accurate time is important (such as trading platforms and financial applications), but they are prohibitively expensive and not a viable solution for those unable to manage their own physical server.
```
Comment (albertpastrana) [inline _drafts/id-generation.md]: More than "those unable to manage their own physical server" I would probably write "those we don't manage their own physical server".

File: _drafts/id-generation.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+* A is the reserved signed bit (which we chose not to use because some platforms make it difficult to get at).
+
+* B is the timestamp in milliseconds since a custom epoch, 41 bits in total.
+
+* C is the logical shard ID, 10 bits in total.
+
+* D is the sequence, 12 bits in total.
+
+We have represented the time as a 41-bit integer by creating a custom epoch: this simply involves picking an arbitrary point in time, calling it "0", and starting to count the milliseconds from there.
```
Comment (albertpastrana) [inline _drafts/id-generation.md]: Maybe we should say here how many time do we have? (~69y)

File: _drafts/id-generation.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+With all this together, once a list of IDs is sorted, they'll be grouped together by time and sorted by the (relatively) meaningless shard ID and sequence number after that. That is, they'll be _k_-sorted by time.
+
+## Redis and Lua
+
+The final piece of the puzzle is in how we use Redis to do this.
+
+Redis lacks the ability to run commands in a transaction based on logic, so we chose to implement the logic in Lua and pass it to Redis using its scripting functionality. At the time of coding, the bitops library provided by Redis didn't allow us to do what we needed to do to compose the 64-bit ID in pure Lua, so we currently call to Redis with the Lua script to get the components which make up the ID (time, logical shard ID, sequence) and bit-shift them all together in the client-library (which is currently written in Java).
+
```
Comment (albertpastrana) [inline _drafts/id-generation.md]: I would add here pseudo-code showing how we do it.

**Pull request merged**